contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp

```diff
@@ -267,6 +319,493 @@ class AlembicScene::AlembicReader : public AlembicIO
+
+			const AbcA::PropertyHeader *propertyHeader = userProperties.getPropertyHeader( name.string() );
+
+			return propertyHeader != nullptr;
+		}
+
+		void attributeNames( NameList &attrs ) const
+		{
+			attrs.clear();
+
+			if( !m_xform )
+			{
+				return;
+			}
+
+			ICompoundProperty userProperties = m_xform.getSchema().getUserProperties();
+
+			if( !userProperties.valid() )
+			{
+				return;
+			}
+
+			attrs.reserve( userProperties.getNumProperties() );
+
+			for( size_t i = 0; i < userProperties.getNumProperties(); ++i )
+			{
+				attrs.push_back( InternedString( userProperties.getPropertyHeader( i ).getName() ) );
+			}
+		}
+
+		size_t numAttributeSamples( const Name &name ) const
+		{
+			if( !m_xform )
+			{
+				return 0;
+			}
+
+			ICompoundProperty userProperties = m_xform.getSchema().getUserProperties();
+
+			const PropertyHeader *header = userProperties.getPropertyHeader( name.string() );
+
+			if( !header )
+			{
+				return 0;
+			}
+
+			return header->getTimeSampling()->getNumStoredTimes();
+		}
+
+		double attributeSampleTime( const Name &name, size_t sampleIndex ) const
+		{
+			if( !m_xform )
+			{
+				return 0.0;
+			}
+
+			ICompoundProperty userProperties = m_xform.getSchema().getUserProperties();
+			double t = userProperties.getPropertyHeader( name.string() )->getTimeSampling()->getSampleTime( sampleIndex );
+			return t;
+		}
+
+		double attributeSampleInterval( const Name &name, double time, size_t &floorIndex, size_t &ceilIndex ) const
+		{
+			if( !m_xform )
+			{
+				return 0.0;
+			}
+
+			ICompoundProperty userProperties = m_xform.getSchema().getUserProperties();
+			const PropertyHeader *propertyHeader = userProperties.getPropertyHeader( name.string() );
+
+			if( !propertyHeader )
+			{
+				return 0.0;
+			}
+
+			auto timeSampling = propertyHeader->getTimeSampling();
+
+			return sampleInterval( timeSampling.get(), timeSampling->getNumStoredTimes(), time, floorIndex, ceilIndex );
+		}
+
+		IECore::ConstObjectPtr readAttributeAtSample( const Name &name, size_t sampleIndex ) const
+		{
+			if( !m_xform )
+			{
+				return nullptr;
+			}
+
+			const IXformSchema &schema = m_xform.getSchema();
+			auto userProperties = schema.getUserProperties();
+
+			const PropertyHeader *propertyHeader = userProperties.getPropertyHeader( name.string() );
+			if( !propertyHeader )
+			{
+				return nullptr;
+			}
+
+			if( propertyHeader->getPropertyType() != kScalarProperty )
+			{
+				IECore::msg(
+					IECore::Msg::Warning,
+					"AlembicScene::readAttributeAtSample",
+					boost::format( "Unsupported property type :%1%. Only scalar properties are currently supported: %2%" ) %
+						propertyHeader->getPropertyType() %
+						kScalarProperty
+				);
+			}
+
+			CompoundPropertyReaderPtr propertyReader = GetCompoundPropertyReaderPtr( userProperties );
+
+			ScalarPropertyReaderPtr scalarPropertyReader = propertyReader->getScalarProperty( name.string() );
+			DataType dataType = scalarPropertyReader->getDataType();
```

Comment:
Assigning these to a variable forces a reader to wonder if they are going to be modified, and then look through the rest of this 400 line function to check that they're not. Is this an attempt at improving clarity or is it important for performance? If for clarity, I would suggest that the variable isn't really helping, at least for me. If it's for performance (or if you strongly prefer them anyway), could you make them const so it's obvious they won't change?

Reply:
>  Is this an attempt at improving clarity or is it important for performance?

Attempt to reduce line length. I started with reading those values when required but refactored to this unclear state. Perhaps I should reduce the size of this function, it's a bit of a beast.